### PR TITLE
Re-enable `BackTracking`

### PR DESCRIPTION
**Pull request by SouthEndMusic**
`BackTracking` as relaxation is now enabled again, with a thin wrapper to reject it when the residual gets worse. Upstream issue: 

https://github.com/SciML/OrdinaryDiffEq.jl/issues/2442

### CODE REVIEW

**Comment by visr** [comment]
Since the sensible gradients don't fix the Crystal model with BackTracking, I guess we should not enable that yet?

How do these gradients compare with Shampine's advice in https://github.com/Deltares/Ribasim/issues/1705#issuecomment-2283312301, should we bring that back as well in this PR?

**Comment by SouthEndMusic** [comment]
I added Shampine's advice, but it doesn't seem to change anything

File: core/src/solve.jl [hunk 1]
```diff
@@ -51,9 +51,20 @@ function water_balance!(
     # Formulate du (controlled by PidControl)
     formulate_du_pid_controlled!(du, graph, pid_control)
 
+    # https://github.com/Deltares/Ribasim/issues/1705#issuecomment-2283293974
+    adapt_negative_storage_du!(du, u)
+
     return nothing
 end
 
+function adapt_negative_storage_du!(du, u)
```
Comment (visr) [inline core/src/solve.jl:60]:
```suggestion
    stop_declining_negative_storage!(du, u)

    return nothing
end

function stop_declining_negative_storage!(du, u)
```

File: core/src/util.jl [hunk 2]
```diff
@@ -887,3 +896,66 @@ end
...
+get_area_and_level(basin::Basin, state_idx::Int, storage::GradientTracer) = storage, storage
+adapt_negative_storage_du!(du, u::ComponentVector{<:GradientTracer}) = nothing
+
+@kwdef struct MonitoredBackTracking{B, V}
+    linesearch::B = BackTracking()
+    dz_tmp::V = []
+    z_tmp::V = []
+end
+
+function resid(z, integrator, nlsolver, f)
```
Comment (visr) [inline core/src/util.jl:908]: Needs a docstring.

Comment (evetion) [reply core/src/util.jl:908]: Or a rename, now I think of it as res_id, not residual. 🤷🏻‍♂️

File: core/src/config.jl [hunk 3]
```diff
@@ -230,7 +230,7 @@ const algorithms = Dict{String, Type}(
 )
 
 "Create an OrdinaryDiffEqAlgorithm from solver config"
-function algorithm(solver::Solver)::OrdinaryDiffEqAlgorithm
+function algorithm(solver::Solver, u0 = [])::OrdinaryDiffEqAlgorithm
```
[the remark this answers is not in the document]
Comment (SouthEndMusic) [reply core/src/config.jl:233]: The default value is only there to not have to specify `u0` in tests where it's not relevant.

**Review by visr — APPROVED** [review]
Very nice find!

**Pull request merged**